core/ipc/sink/aws-sqs-impl/src/main/java/org/opennms/core/ipc/sink/aws/sqs/AwsMessageConsumerManager.java

```diff
@@ -0,0 +1,214 @@
+
+    /** The Constant LOG. */
+    private static final Logger LOG = LoggerFactory.getLogger(AwsMessageConsumerManager.class);
+
+    /** The consumer runners by module. */
+    private final Map<SinkModule<?, Message>, List<AwsConsumerRunner>> consumerRunnersByModule = new ConcurrentHashMap<>();
+
+    /** The thread factory. */
+    private final ThreadFactory threadFactory = new ThreadFactoryBuilder()
+            .setNameFormat("aws-consumer-%d")
+            .build();
+
+    /** The executor. */
+    private final ExecutorService executor = Executors.newCachedThreadPool(threadFactory);
+
+    /** The AWS configuration. */
+    private final Properties awsConfig = new Properties();
+
+    /**
+     * The Class AwsConsumerRunner.
+     */
+    private class AwsConsumerRunner implements Runnable {
+
+        /** The module. */
+        private final SinkModule<?, Message> module;
+
+        /** The closed. */
+        private final AtomicBoolean closed = new AtomicBoolean(false);
+
+        /** The consumer. */
+        private final MessageConsumer consumer;
+
+        /** The connection. */
+        private final SQSConnection connection;
+
+        /**
+         * Instantiates a new AWS consumer runner.
+         *
+         * @param module the module
+         * @throws JMSException the JMS exception
+         */
+        public AwsConsumerRunner(SinkModule<?, Message> module) throws JMSException {
+            this.module = module;
+            final String queueName = AwsUtils.getQueueName(module);
+            connection = AwsUtils.createConnection(awsConfig);
+            AwsUtils.ensureQueueExists(connection, queueName);
+            final Session session = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+            consumer = session.createConsumer(session.createQueue(queueName));
+        }
+
+        /* (non-Javadoc)
+         * @see java.lang.Runnable#run()
+         */
+        @Override
+        public void run() {
+            Logging.putPrefix(MessageConsumerManager.LOG_PREFIX);
+            try {
+                connection.start();
+                while (!closed.get()) {
+                    javax.jms.Message message = consumer.receive(30000);
+                    LOG.debug("Got message {}", message.getJMSMessageID());
+                    if (message instanceof TextMessage) {
+                        TextMessage txtMessage = (TextMessage) message;
+                        try {
+                            dispatch(module, module.unmarshal(txtMessage.getText()));
+                        } catch (RuntimeException e) {
+                            LOG.warn("Unexpected exception while dispatching message", e);
+                        }
+                    } else {
+                        LOG.warn("Received an unsupported message {}", message.getJMSMessageID());
+                    }
+                }
+            } catch (JMSException ex) {
+                LOG.error("Can't receive message", ex);
```

Comment:
How do we recover from these errors? If the consumer fails for some reason, possibly a network interruption, we likely don't want to just exit the thread and stop consuming altogether.

Reply:
I understand. The reason why I've added the "try" at that level is because `connection.start();` can throw a JMSException, which can be for example due to bad credentials, on those extremely rare occasions where SQS is not available on the chosen region. What do you suggest to do there ? Maybe just do the try/catch for `connection.start`, and if it goes well, enter the receive loop, if not leave the thread ?

Here is what the documentation says about the `connect` method:

```
You can call the start method on the connection at any time.
However, the consumer does not begin to receive messages until you call it.
```